jobbergate-core/jobbergate_core/auth/core.py

```diff
@@ -0,0 +1,316 @@
+"""
+Utilities for handling authentication in the Jobbergate system.
+"""
+import time
+from collections import namedtuple
+from dataclasses import dataclass, field
+from pathlib import Path
+from typing import Dict, cast
+
+import httpx
+from loguru import logger
+
+from .exceptions import AuthenticationError, TokenError
+from .token import Token, TokenType
+
+
+_LoginInformation = namedtuple(
+    "_LoginInformation",
+    ["verification_url", "wait_interval", "device_code"],
+)
+
+
+@dataclass
+class JobbergateAuth:
+    """
+    High-level class used to handle authentication in Jobbergate.
+
+    After an instance of this class is created, it can be used to authenticate requests
+    from both `requests`_  and `httpx`_ packages by passing it to the ``auth``
+    parameter on the request (see examples bellow).
+
+    It works just out of the box. Behind the scenes, this procedure calls the
+    :meth:`JobbergateAuth.acquire_tokens` method to load the available tokens from the cache directory, it
+    will try to refresh them if they are expired, or will provide an URL to the user to login on the system.
+
+    Notice all steps above are also available individually as public methods, allowing a fine control for
+    advanced users.
+
+    .. _requests: https://requests.readthedocs.io/en/latest/
+    .. _httpx: https://www.python-httpx.org/
+
+    Arguments:
+        cache_directory (Path): Directory to be used for the caching tokens.
+        login_domain (str): Domain used for the login.
+        login_audience (str): Audience of the login.
+        login_client_id (str): Client ID used for login.
+        tokens (Dict[TokenType, Token]): Dictionary holding the tokens needed for authentication.
```

Comment:
I don't love the type hints in the docstring since they are already defined in the class. This seems like it could be a synchronization problem if the type of one of these attributes changes but updating the docs is forgotten.

Reply:
See the image below:
![image](https://user-images.githubusercontent.com/45343359/222430601-2785e602-ba7b-4d90-823d-76d504799996.png)

Visually saying, it is very good IMO to have the type hints in the docstring. The following example doesn't have the type hints:
![image](https://user-images.githubusercontent.com/45343359/222430772-1d1c6056-2822-41b9-8e9c-1d678fd365dc.png)

For me, the way Felipe did is far better because it helps people with any sort of visual disability. Specially when there are many parameters, trying to find the type hint in the second example is terrible IMO. See the following for example:
![image](https://user-images.githubusercontent.com/45343359/222431522-c260200b-79c2-4550-a611-32a25ad479a6.png)

Reply:
I agree with both. On internal methods, I agree this is just duplication. On the other hand, I see value here where we are exposing them on the documentation, as Tosta pointed out.

Can we agree on removing them on internal details and keeping them on what is on the documentation? @dusktreader  @matheushent

Reply:
> Can we agree on removing them on internal details and keeping them on what is on the documentation?
I didn't understand, sorry. Did you mean to keep as is?

Reply:
agreed

Reply:
There is still a synchronization problem. Any time you have to remember to update documentation when you change code is an opportunity to have docs go out of date. For example you could leave the functionality of a function the same such that you think you don't need to change the documentation about what the function does. However, if you change the type of an argument, you have to remember to update the docstring as well. This sort of 'drift' between implementation and documentation can be a big problem.

I will resolve this comment for now. However, I would like to see if we can find a tool that knows how to incorporate type-hints into parameter docs without requiring the user to annotate it in the docstring.